versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/spi/AbstractDatabaseAdapter.java

```diff
@@ -2091,8 +2091,9 @@ protected CommitLogEntry squashCommits(
       CommitLogEntry source = commitsToMergeChronological.get(i);
       for (ContentKey delete : source.getDeletes()) {
         if (includeKeyPredicate.test(delete)) {
-          deletes.add(delete);
-          puts.remove(delete);
+          if (puts.remove(delete) == null) {
+            deletes.add(delete);
```

Comment:
What about a `Put/Delete` pair in the same (single) commit? It may be meaningful when content IDs in the deleted and added objects differ.

Reply:
I suppose we ought to retain both the `Delete` and the `Put` operations in this case (according to the latest OpenAPI descriptions updates in #6438).

Would you mind adding a test case to see how it works?

Reply:
Hmm how can you achieve that? I tried:

1) DELETE and PUT in the same source commit:

```java
    commit("C1").put(key1, V_1_1).put(key2, V_2_1).put(key3, V_3_1).toBranch(source);
    commit("C2").delete(key1).delete(key2).delete(key3).put(key3, V_3_1.withId("changed")).toBranch(source);
    Hash sourceHead = commit("C3").put(key2, V_2_2).toBranch(source);
```

This gives me the error (new model):

```
java.lang.IllegalArgumentException: Duplicate key in commit operations: t3
```

2) DELETE and PUT in two subsequent source commits, with a manually assigned content ID:

```java
    commit("C1").put(key1, V_1_1).put(key2, V_2_1).put(key3, V_3_1).toBranch(source);
    commit("C2").delete(key1).delete(key2).delete(key3).toBranch(source);
    Hash sourceHead = commit("C3").put(key2, V_2_2).put(key3, V_3_1.withId("changed")).toBranch(source);
```

This gives me the error (new model):

```
java.lang.IllegalArgumentException: New value for key 't3' must not have a content ID
```

Reply:
Re: Snipped 1: perhaps the old model does not follow (or cannot follow?) the latest API spec?

Reply:
Oooh are you referring to this?

> "If the key for a content shall change (aka a rename), then use a `Delete` "
              + "operation using the current (old) key and a `Put` operation using the new key "
              + "with the `value` having the correct content-id. Both operations must happen "
              + "in the same commit.\n"

I think this is the opposite case: it's a rename. I can add a test case for that.

Reply:
@dimas-b I added a rename test case in 294345c.

Reply:
That's a good test/use case to keep :+1: 

However, I meant this:
```
              + "A content object can be replaced (think: `DROP TABLE xyz` + `CREATE TABLE xyz`) "
              + "with a `Delete` operation and a `Put` operation for a content using a `value`"
              + "representing a new content object, so without a content-id, in the same commit.")
```

Reply:
Ah that's clearly not possible, neither in the old nor the new model, regardless of the changes in this PR. The PUT must follow the DELETE in a separate commit.

I think we should amend the API specs, wdyt?

Reply:
Yeah, let's fix the API spec (in a separate PR maybe)?

Reply:
The `Duplicate key in commit operations` error looks like a bug in the new model to me (`CommitImpl`). I think API permits `Put/Delete` pairs for the same key in a single commit.

Reply:
Looking at the current API spec/docs I think we should actually keep `Put/Delete` pairs in the squashed commit, just maybe use the most recent `Put` in this case. WDYT?

Reply:
> The Duplicate key in commit operations error looks like a bug in the new model to me (CommitImpl). I think API permits Put/Delete pairs for the same key in a single commit.

The old model does not allow it either, see [here](https://github.com/projectnessie/nessie/blob/nessie-0.54.0/versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/spi/AbstractDatabaseAdapter.java#L322-L327).

Looking at the Git history of this file, we see that this check was added in #2311. The PR description by @snazy is enlightening:

    While thinking about how Nessie could actually squash multiple commits, I realized
    that this can result in multiple commit-operations for the same content-key.

    While commit-operations are ordered on the (REST) API level (list of Operations),
    database adapters split those into separate collections of delete and put operations,
    which makes it impossible to determine the original order of the squashed operations.

    The Nessie spec does not define the behavior of a Nessie commit with multiple
    operations for the same key in one commit.

    The spec now clearly states that there must be no more than [one] operation (put / delete)
    for each content key.

The above PR introduced [this sentence](https://github.com/projectnessie/nessie/blob/nessie-0.54.0/site/docs/develop/spec.md?plain=1#L144) in `site/docs/develop/spec.md`: "A Content Key must only occur once in a Nessie commit." The sentence is still there.

> Looking at the current API spec/docs I think we should actually keep Put/Delete pairs in the squashed commit, just maybe use the most recent Put in this case. WDYT?

How would we know in which order the operations should be applied? PUT then DELETE, or DELETE then PUT? Allowing the same key to appear more than once would only be feasible if we could determine the order in which operations in the same commit should be applied.

All in all, I think the sentence introduced recently [here](https://github.com/projectnessie/nessie/blob/nessie-0.54.0/api/model/src/main/java/org/projectnessie/model/Operation.java#L75-L77) is wrong and should be amended.

>  It might be worth adding API-level tests for this too (BaseTestNessieApi)

Sure I will add some.

Reply:
There are two scenarios:
1. rename (Delete + Put w/ previous content-ID, must not change payload)
2. re-create (Detlete + Put w/o content-ID, can change payload)

There are test cases on the VersionStore level that validate that though (I _think_ I added tests for these cases).

Yes, the sentence that a single key must only appear once in a commit is correct - except for the rename & re-create cases.